provers/risc0/driver/src/bonsai/auto_scaling.rs

```diff
@@ -0,0 +1,166 @@
+            // Parse the JSON response
+            let data: ScalerResponse = response.json().await.unwrap_or_default();
+            debug!("Scaler status: {:?}", data);
+            Ok(data)
+        } else {
+            trace_err!("Request failed with status: {}", response.status());
+            Err(Error::msg("Failed to get bonsai gpu num".to_string()))
+        }
+    }
+
+    async fn set_bonsai_gpu_num(&self, gpu_num: u32) -> Result<()> {
+        // Create a new client
+        let client = Client::new();
+
+        // Make the POST request
+        let response = client
+            .post(self.url.clone())
+            .headers(self.headers.clone())
+            .body(gpu_num.to_string())
+            .send()
+            .await?;
+
+        // Check if the request was successful
+        if response.status().is_success() {
+            // Parse the JSON response
+            let data: ScalerResponse = response.json().await?;
+
+            // Use the parsed data
+            debug!("Scaler status: {:?}", data);
+            assert_eq!(data.desired, gpu_num);
+            Ok(())
+        } else {
+            trace_err!("Request failed with status: {}", response.status());
+            Err(Error::msg("Failed to get bonsai gpu num".to_string()))
+        }
+    }
+}
+
+lazy_static! {
+    static ref BONSAI_API_URL: String =
+        env::var("BONSAI_API_URL").expect("BONSAI_API_URL must be set");
+    static ref BONSAI_API_KEY: String =
+        env::var("BONSAI_API_KEY").expect("BONSAI_API_KEY must be set");
+}
+
+const MAX_BONSAI_GPU_NUM: u32 = 15;
+
+pub(crate) async fn maxpower_bonsai() -> Result<()> {
+    let auto_scaler = BonsaiAutoScaler::new(BONSAI_API_URL.to_string(), BONSAI_API_KEY.to_string());
+    let current_gpu_num = auto_scaler.get_bonsai_gpu_num().await?;
+    // either already maxed out or pending to be maxed out
+    if current_gpu_num.current == MAX_BONSAI_GPU_NUM
+        || (current_gpu_num.current + current_gpu_num.pending == MAX_BONSAI_GPU_NUM)
+    {
+        Ok(())
+    } else {
+        auto_scaler.set_bonsai_gpu_num(MAX_BONSAI_GPU_NUM).await?;
+        // wait for the bonsai to heat up
+        tokio::time::sleep(tokio::time::Duration::from_secs(180)).await;
+        let scaler_status = auto_scaler.get_bonsai_gpu_num().await?;
+        assert!(scaler_status.current == MAX_BONSAI_GPU_NUM);
+        Ok(())
+    }
+}
+
+pub(crate) async fn shutdown_bonsai() -> Result<()> {
+    let auto_scaler = BonsaiAutoScaler::new(BONSAI_API_URL.to_string(), BONSAI_API_KEY.to_string());
+    let current_gpu_num = auto_scaler.get_bonsai_gpu_num().await?;
+    if current_gpu_num.current == 0 {
+        Ok(())
+    } else {
+        auto_scaler.set_bonsai_gpu_num(0).await?;
+        // wait few minute for the bonsai to cool down
+        tokio::time::sleep(tokio::time::Duration::from_secs(60)).await;
+        let scaler_status = auto_scaler.get_bonsai_gpu_num().await?;
+        assert!(scaler_status.current == 0);
```

Comment:
Similar to the above comment. This makes the awaiting for the result reusable, and we just use the result to handle the logic of the asserts outside.